plat/common/include/x86/cpu.h

```diff
@@ -97,6 +97,21 @@ static inline __u64 rdtsc(void)
 	return (h << 32) | l;
 }
 
+#ifdef __FSGSBASE__
+static inline __u64 rdfsbase(void)
+{
+	__u64 ret;
+
+	__asm__ __volatile__ ("rdfsbase %%rax" : "=a" (ret));
```

Comment:
Yeah, good point; or maybe even intrinsics, that way it's more compiler-agnostic; patch incoming.

Reply:
It seems that simply adding "-mfsgsbase" in the compilation parameters is sufficient, without the need to include the header file?

Reply:
Yes, for the builtins (for example `__builtin_ia32_rdfsbase64` the parameter is enough). There are also intrinsics in `immintrin.h` (for example `_readfsbase_u64`). But the builtins are documented anyway and shouldn't change.

The feature levels are recent, yes. They are only a shortcut to enable instruction set extensions for a "generation" of processors. Of course, enabling these individually is an option, but the resulting code will still only work on Ivy Bridge generation-ish processors and using the feature level would also enable all the other nice to have instructions (AVX, movbe, etc.).

Reply:
Do these `__builtin_ia32_rdfsbase64` depend on to compile and optimize for a certain CPU? IT should be better independent so that we can still run on a wider range of x86 CPU generations. The right way is probably cpuid, right?